gettsim/taxes/zu_verst_eink/vorsorge.py

```diff
@@ -0,0 +1,230 @@
+    Contributions to other security branches can also be deducted.
+
+    This ruling differs to vorsorge2005() only in the treatment of other contributions.
+
+    Parameters
+    ----------
+    _altervorsorge_aufwend
+    pflegev_beit_m
+    ges_krankenv_beitr_m
+    arbeitsl_v_beit_m
+    eink_st_abzuege_params
+
+    Returns
+    -------
+
+    """
+    out = copy.deepcopy(_altervorsorge_aufwend) * 0
+    # 'Basisvorsorge': Health and old-age care contributions are deducted anyway.
+    sonstige_vors = 12 * (
+        pflegev_beit_m.loc[~kind]
+        + (1 - eink_st_abzuege_params["vorsorge_kranken_minderung"])
+        * ges_krankenv_beitr_m.loc[~kind]
+    )
+    # maybe add unemployment insurance, but do not exceed 1900€.
+    out.loc[~kind] = sonstige_vors.clip(
+        lower=(sonstige_vors + 12 * arbeitsl_v_beit_m.loc[~kind]).clip(
+            upper=eink_st_abzuege_params["vorsorge_sonstige_aufw_max"]
+        )
+    )
+    out.loc[~kind] += _altervorsorge_aufwend.loc[~kind]
+    return out.rename("_vorsorge_alternative_ab_2010")
+
+
+def vorsorge_bis_2004(
+    _lohn_vorsorgeabzug_single,
+    _lohn_vorsorgeabzug_tu,
+    ges_krankenv_beitr_m,
+    rentenv_beit_m,
+    ges_krankenv_beitr_m_tu,
+    rentenv_beitr_m_tu,
+    tu_id,
+    gemeinsam_veranlagte_tu,
+    gem_veranlagt,
+    kind,
+    eink_st_abzuege_params,
+):
+    out = copy.deepcopy(ges_krankenv_beitr_m) * 0
+    out.loc[~gem_veranlagt & ~kind] = berechne_vorsorge_bis_2004(
+        _lohn_vorsorgeabzug_single.loc[~kind],
+        ges_krankenv_beitr_m.loc[~gem_veranlagt & ~kind],
+        rentenv_beit_m.loc[~gem_veranlagt & ~kind],
+        1,
+        eink_st_abzuege_params,
+    )
+    vorsorge_tu = berechne_vorsorge_bis_2004(
+        _lohn_vorsorgeabzug_tu,
+        ges_krankenv_beitr_m_tu.loc[gemeinsam_veranlagte_tu],
+        rentenv_beitr_m_tu.loc[gemeinsam_veranlagte_tu],
+        2,
+        eink_st_abzuege_params,
+    )
+    out.loc[gem_veranlagt & ~kind] = tu_id[gem_veranlagt].replace(vorsorge_tu)
+    return out.rename("vorsorge")
+
+
+def _lohn_vorsorge_bis_2019_single(
+    bruttolohn_m, gemeinsam_veranlagt, eink_st_abzuege_params
+):
+    out = (
+        eink_st_abzuege_params["vorsorge2004_vorwegabzug"]
+        - eink_st_abzuege_params["vorsorge2004_kürzung_vorwegabzug"]
+        * 12
+        * bruttolohn_m.loc[~gemeinsam_veranlagt]
+    ).clip(lower=0)
+    return out.rename("_lohn_vorsorge_bis_2019_single")
+
+
+def _lohn_vorsorgeabzug_bis_2019_tu(
+    bruttolohn_m_tu, gemeinsam_veranlagte_tu, eink_st_abzuege_params
+):
+    out = 0.5 * (
+        2 * eink_st_abzuege_params["vorsorge2004_vorwegabzug"]
+        - eink_st_abzuege_params["vorsorge2004_kürzung_vorwegabzug"]
+        * 12
+        * bruttolohn_m_tu.loc[gemeinsam_veranlagte_tu]
+    ).clip(lower=0)
+    return out.rename("_lohn_vorsorgeabzug_bis_2019_tu")
+
+
+def _lohn_vorsorge_ab_2020_single(gemeinsam_veranlagt):
+    out = gemeinsam_veranlagt.loc[gemeinsam_veranlagt].astype(int) * 0
```

Comment:
what's causing the 2020/2019 difference? This affects only the 2004 regime right? This does not become clear from the function name

Reply:
It's only for the 2004 regime. The amount for the old reform til 2004 and today's rule and select the better solution for the individual? Will this expire in 2020?

Reply:
After a discussion on Zulip: From 2020 there is no check with the earlier reform. Therefore I deleted the functions from 2020 onwards and assigned directly the `vorsorge` function!